--- src/generate.c
+++ src/generate.c
@@ -6097,8 +6097,13 @@
	bool lit_level = FALSE;

	byte level_bg = FEAT_WALL_EXTRA;
-	int base_rooms = (p_ptr->depth < 10) ? DUN_ROOMS : (DUN_ROOMS + p_ptr->depth * 2);
+
+	int base_rooms = (p_ptr->depth < 10) ? DUN_ROOMS : (DUN_ROOMS + p_ptr->depth);
+	if (base_rooms > (DUN_ROOMS * 2)) {
+		base_rooms = DUN_ROOMS * 2;
+	}
+
	s16b dun_rooms = base_rooms * 2;

	/* Global data */
